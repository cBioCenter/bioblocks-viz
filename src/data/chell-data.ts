// Type definitions for Chell Visualization Library
// Project: https://github.com/cBioCenter/chell-viz
// Definitions by: cBioCenter @ Dana-Farber <https://github.com/cBioCenter/>, James Lindsay <https://github.com/jim-bo>, Nick Gauthier <https://github.com/npgauth>, Drew Diamantoukos <https://github.com/MercifulCode>,
// TypeScript Version: 2.8
import { Structure } from 'ngl';
import { ISpringGraphData } from 'spring';
import Chell1DSection from './Chell1DSection';
import { ChellPDB } from './ChellPDB';
import { CouplingContainer } from './CouplingContainer';

export type CONTACT_MAP_DATA_TYPE = IContactMapData;
export type NGL_DATA_TYPE = Structure;
export type SPRING_DATA_TYPE = ISpringGraphData;
export type T_SNE_DATA_TYPE = number[][];

export type CHELL_DATA_TYPE = CONTACT_MAP_DATA_TYPE | NGL_DATA_TYPE | SPRING_DATA_TYPE | T_SNE_DATA_TYPE;

// TODO Better define Cell type to not be indices.
/** Shared data between Spring/T-SNE. Refers to index in points collection. */
export type CELL_TYPE = number;
export type RESIDUE_TYPE = number;

export enum VIZ_TYPE {
  CONTACT_MAP = 'Contact Map',
  NGL = 'NGL',
  SPRING = 'Spring',
  'T-SNE' = 'T-SNE',
}

export interface IContactMapData {
  couplingScores: CouplingContainer;
  pdbData?: ChellPDB;
  secondaryStructures: ISecondaryStructureData[];
}

export interface IMonomerContact {
  i: number;
  j: number;
  dist: number;
}

export interface ICouplingScore {
  i: number;
  A_i?: string;
  j: number;
  A_j?: string;
  fn?: number;
  cn?: number;
  segment_i?: string;
  segment_j?: string;
  probability?: number;
  dist_intra?: number;
  dist_multimer?: number;
  dist: number;
  precision?: number;
}

export enum CONTACT_DISTANCE_PROXIMITY {
  CLOSEST = 'CLOSEST',
  C_ALPHA = 'C-Î±',
}

export enum CONFIGURATION_COMPONENT_TYPE {
  RADIO = 'RADIO',
  SLIDER = 'SLIDER',
}

export interface ISecondaryStructureData {
  resno: number;
  structId: SECONDARY_STRUCTURE_KEYS;
}

export type SECONDARY_STRUCTURE_SECTION = Chell1DSection<SECONDARY_STRUCTURE_KEYS>;
export type SECONDARY_STRUCTURE = Array<Chell1DSection<SECONDARY_STRUCTURE_KEYS>>;

export interface IResiduePair {
  i: number;
  j: number;
}

export enum SECONDARY_STRUCTURE_CODES {
  'G' = '310_HELIX',
  'H' = 'ALPHA_HELIX',
  'I' = 'PI_HELIX',
  'T' = 'HYDROGEN_BONDED_TURN',
  'E' = 'BETA_SHEET',
  'B' = 'BETA_BRIDGE',
  'S' = 'BEND',
  'C' = 'COIL',
}

export type SECONDARY_STRUCTURE_KEYS = keyof typeof SECONDARY_STRUCTURE_CODES;

export interface ISecondaryStructure {
  code: SECONDARY_STRUCTURE_CODES;
  name: string;
  minResidueLength: number;
}
