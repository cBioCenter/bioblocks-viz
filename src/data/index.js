"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
tslib_1.__exportStar(require("./AminoAcid"), exports);
tslib_1.__exportStar(require("./Nucleotide"), exports);
tslib_1.__exportStar(require("./AnatomogramMapping"), exports);
tslib_1.__exportStar(require("./Bioblocks1DSection"), exports);
tslib_1.__exportStar(require("./BioblocksConfig"), exports);
tslib_1.__exportStar(require("./BioblocksData"), exports);
tslib_1.__exportStar(require("./BioblocksPDB"), exports);
tslib_1.__exportStar(require("./CodonTable"), exports);
tslib_1.__exportStar(require("./CouplingContainer"), exports);
tslib_1.__exportStar(require("./Dataset"), exports);
tslib_1.__exportStar(require("./IUPACData"), exports);
tslib_1.__exportStar(require("./Markers"), exports);
tslib_1.__exportStar(require("./PfamProtein"), exports);
tslib_1.__exportStar(require("./PlotlyData"), exports);
tslib_1.__exportStar(require("./Protein"), exports);
tslib_1.__exportStar(require("./Seq"), exports);
tslib_1.__exportStar(require("./SeqIO"), exports);
tslib_1.__exportStar(require("./SeqRecord"), exports);
tslib_1.__exportStar(require("./TintedBioblocks1DSection"), exports);
tslib_1.__exportStar(require("./event"), exports);
//# sourceMappingURL=index.js.map