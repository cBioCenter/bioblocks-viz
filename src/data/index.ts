export * from './AminoAcid';
export * from './Chell1DSection';
export * from './ChellConfig';
export * from './ChellData';
export * from './ChellPDB';
export * from './CouplingContainer';
export * from './PlotlyData';
export * from './Protein';
export * from './Spring';
export * from './TintedChell1DSection';

export * from './event';
