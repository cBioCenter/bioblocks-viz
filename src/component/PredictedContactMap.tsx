import * as React from 'react';
import { initialResidueContext } from '../context/ResidueContext';
import { CONFIGURATION_COMPONENT_TYPE, CONTACT_DISTANCE_PROXIMITY, IContactMapData } from '../data/chell-data';
import { CouplingContainer } from '../data/CouplingContainer';
import { withDefaultProps } from '../helper/ReactHelper';
import { generateChartDataEntry, IContactMapChartData } from './chart/ContactMapChart';
import ContactMap, { IContactMapConfiguration } from './ContactMap';

export const defaultPredictedContactMapProps = {
  correctColor: '#ff0000',
  data: {
    couplingScores: new CouplingContainer(),
  } as IContactMapData,
  height: 400,
  incorrectColor: '#000000',
  ...initialResidueContext,
  observedColor: '#0000ff',
  padding: 0,
  width: 400,
};

export const initialPredictedContactMapState = {
  chainLength: -1,
  linearDistFilter: 5,
  measuredContactDistFilter: 5,
  measuredProximity: CONTACT_DISTANCE_PROXIMITY.CLOSEST,
  numPredictionsToShow: -1,
  pointsToPlot: [] as IContactMapChartData[],
};

export type PredictedContactMapProps = {} & typeof defaultPredictedContactMapProps;
export type PredictedContactMapState = Readonly<typeof initialPredictedContactMapState>;

export class PredictedContactMapClass extends React.Component<PredictedContactMapProps, PredictedContactMapState> {
  public readonly state: PredictedContactMapState = initialPredictedContactMapState;

  constructor(props: PredictedContactMapProps) {
    super(props);
  }

  public onLinearDistFilterChange = () => (value: number) => {
    this.setState({
      linearDistFilter: value,
    });
  };

  public onNumPredictionsToShowChange = () => (value: number) => {
    this.setState({
      numPredictionsToShow: value,
    });
  };

  public onMeasuredProximityChange = () => (value: number) => {
    this.setState({
      measuredProximity: Object.values(CONTACT_DISTANCE_PROXIMITY)[value],
    });
  };

  public componentDidUpdate(prevProps: PredictedContactMapProps, prevState: PredictedContactMapState) {
    const { correctColor, data, incorrectColor, observedColor } = this.props;
    const { linearDistFilter, measuredContactDistFilter, measuredProximity, numPredictionsToShow } = this.state;

    const isRecomputeNeeded =
      data.couplingScores !== prevProps.data.couplingScores ||
      linearDistFilter !== prevState.linearDistFilter ||
      measuredContactDistFilter !== prevState.measuredContactDistFilter ||
      measuredProximity !== prevState.measuredProximity ||
      numPredictionsToShow !== prevState.numPredictionsToShow;
    if (isRecomputeNeeded) {
      const couplingScores = data.pdbData
        ? data.pdbData.generateCouplingsAmendedWithPDB(data.couplingScores.rankedContacts, measuredProximity)
        : new CouplingContainer(data.couplingScores.rankedContacts);

      const { chainLength } = couplingScores;
      const observedContacts = couplingScores.getObservedContacts(measuredContactDistFilter, linearDistFilter);

      const allPredictions = couplingScores.getPredictedContacts(numPredictionsToShow, linearDistFilter);

      const correctPredictionPercent = (
        (allPredictions.correct.length / allPredictions.predicted.length) *
        100
      ).toFixed(2);
      const newPoints = [
        generateChartDataEntry(
          'x+y',
          { start: observedColor, end: 'rgb(100,177,200)' },
          'Known Structure Contact',
          '(from PDB structure)',
          4,
          observedContacts,
        ),
        generateChartDataEntry(
          'x+y',
          incorrectColor,
          'Predicted Contact',
          `(N=${numPredictionsToShow}, L=${chainLength})`,
          4,
          allPredictions.predicted,
        ),
        generateChartDataEntry(
          'x+y',
          correctColor,
          'Correct Prediction',
          `(N=${allPredictions.correct.length}, ${correctPredictionPercent}%)`,
          6,
          allPredictions.correct,
        ),
      ] as IContactMapChartData[];

      this.setState({
        chainLength,
        numPredictionsToShow:
          data.couplingScores !== prevProps.data.couplingScores ? Math.floor(chainLength / 2) : numPredictionsToShow,
        pointsToPlot: newPoints,
      });
    }
  }

  public render() {
    const { data, ...passThroughProps } = this.props;
    const { chainLength, pointsToPlot } = this.state;
    return (
      <div id="PredictedContactMapComponent">
        <ContactMap
          chainLength={chainLength}
          configurations={this.getContactMapConfigs()}
          data={{
            computedPoints: pointsToPlot,
            secondaryStructures: data.pdbData ? data.pdbData.secondaryStructureSections : [],
          }}
          {...passThroughProps}
        />
      </div>
    );
  }

  protected getContactMapConfigs = (): IContactMapConfiguration[] => [
    {
      name: 'Calculate Distance: Closest Atom - C-Alpha',
      onChange: this.onMeasuredProximityChange(),
      type: CONFIGURATION_COMPONENT_TYPE.RADIO,
      values: {
        current: Object.keys(CONTACT_DISTANCE_PROXIMITY).indexOf(this.state.measuredProximity),
        max: 1,
        min: 0,
        options: Object.values(CONTACT_DISTANCE_PROXIMITY),
      },
    },
    {
      name: 'Linear Distance Filter (|i - j|)',
      onChange: this.onLinearDistFilterChange(),
      type: CONFIGURATION_COMPONENT_TYPE.SLIDER,
      values: {
        current: this.state.linearDistFilter,
        max: 10,
        min: 1,
      },
    },
    {
      name: 'Top N Predictions to Show',
      onChange: this.onNumPredictionsToShowChange(),
      type: CONFIGURATION_COMPONENT_TYPE.SLIDER,
      values: {
        current: this.state.numPredictionsToShow,
        max: this.state.chainLength,
        min: 1,
      },
    },
  ];
}

export const PredictedContactMap = withDefaultProps(defaultPredictedContactMapProps, PredictedContactMapClass);

export default PredictedContactMap;
