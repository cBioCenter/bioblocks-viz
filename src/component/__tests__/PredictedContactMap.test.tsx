import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import * as React from 'react';

import { CONTACT_DISTANCE_PROXIMITY, ICouplingScore, SECONDARY_STRUCTURE_CODES } from '../../data/chell-data';
import { ChellPDB } from '../../data/ChellPDB';
import { CouplingContainer } from '../../data/CouplingContainer';
import { PredictedContactMap, PredictedContactMapClass } from '../PredictedContactMap';

describe('PredictedContactMap', () => {
  const emptyData = {
    couplingScores: new CouplingContainer(),
    secondaryStructures: [],
  };

  const generateCouplingScore = (
    i: number,
    j: number,
    dist: number,
    extra?: Partial<ICouplingScore>,
  ): ICouplingScore => ({
    dist,
    i,
    j,
    ...extra,
  });

  // Translated from example1/coupling_scores.csv
  const sampleCorrectPredictedContacts = [generateCouplingScore(56, 50, 2.4)];
  const sampleIncorrectPredictedContacts = [generateCouplingScore(42, 50, 20.4)];
  const sampleOutOfLinearDistContacts = [
    generateCouplingScore(45, 46, 1.3),
    generateCouplingScore(44, 45, 1.3),
    generateCouplingScore(56, 57, 1.3),
  ];
  const sampleObservedContacts = [...sampleCorrectPredictedContacts, generateCouplingScore(41, 52, 1.3)];

  const uniqueScores = new Set(
    Array.from([
      ...sampleCorrectPredictedContacts,
      ...sampleIncorrectPredictedContacts,
      ...sampleObservedContacts,
      ...sampleOutOfLinearDistContacts,
    ]),
  );

  const sampleData = {
    couplingScores: new CouplingContainer(Array.from(uniqueScores)),
    secondaryStructures: [
      { resno: 30, structId: 'C' as keyof typeof SECONDARY_STRUCTURE_CODES },
      { resno: 31, structId: 'C' as keyof typeof SECONDARY_STRUCTURE_CODES },
    ],
  };

  const sampleDataWithPDB = async () => ({
    couplingScores: new CouplingContainer(Array.from(uniqueScores)),
    pdbData: await ChellPDB.createPDB('sample/protein.pdb'),
    secondaryStructures: [
      { resno: 30, structId: 'C' as keyof typeof SECONDARY_STRUCTURE_CODES },
      { resno: 31, structId: 'C' as keyof typeof SECONDARY_STRUCTURE_CODES },
    ],
  });

  describe('Snapshots', () => {
    it('Should match existing snapshot when given no data.', () => {
      expect(toJson(shallow(<PredictedContactMap />))).toMatchSnapshot();
    });

    it('Should match existing snapshot when given empty data.', () => {
      expect(toJson(shallow(<PredictedContactMap data={emptyData} />))).toMatchSnapshot();
    });

    it('Should match snapshot when locked residues are added.', async () => {
      const wrapper = await shallow(<PredictedContactMap />);
      const expectedSelectedPoints = {
        '37,46': [37, 46],
        '8': [8],
      };
      wrapper.setProps({
        lockedResiduePairs: expectedSelectedPoints,
      });
      await wrapper.update();
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('Sliders', () => {
    it('Should update linear distance filter when appropriate slider is updated.', () => {
      const wrapper = shallow(<PredictedContactMap data={sampleData} />);
      const instance = wrapper.instance() as PredictedContactMapClass;
      const expected = 10;
      expect(instance.state.linearDistFilter).not.toBe(expected);
      instance.onLinearDistFilterChange()(expected);
      instance.forceUpdate();
      expect(instance.state.linearDistFilter).toBe(expected);
    });

    it('Should update number of predicted contacts to show when appropriate slider is updated.', () => {
      const wrapper = shallow(<PredictedContactMap data={sampleData} />);
      const instance = wrapper.instance() as PredictedContactMapClass;
      const expectedCount = 50;
      expect(instance.state.numPredictionsToShow).not.toBe(expectedCount);
      instance.onNumPredictionsToShowChange()(expectedCount);
      wrapper.update();
      expect(instance.state.numPredictionsToShow).toBe(expectedCount);
    });

    it('Should update # of predicted contacts to show when appropriate slider is updated.', () => {
      const wrapper = shallow(<PredictedContactMap data={sampleData} />);
      const instance = wrapper.instance() as PredictedContactMapClass;
      const expected = 20;
      expect(instance.state.numPredictionsToShow).not.toBe(expected);
      instance.onNumPredictionsToShowChange()(expected);
      wrapper.update();
      expect(instance.state.numPredictionsToShow).toBe(expected);
    });

    it('Should update how Measured Proximity is determined.', () => {
      const wrapper = shallow(<PredictedContactMap data={sampleData} />);
      const instance = wrapper.instance() as PredictedContactMapClass;
      const expected = CONTACT_DISTANCE_PROXIMITY.C_ALPHA;
      expect(instance.state.measuredProximity).not.toBe(expected);
      instance.onMeasuredProximityChange()(Object.values(CONTACT_DISTANCE_PROXIMITY).indexOf(expected));
      wrapper.update();
      expect(instance.state.measuredProximity).toBe(expected);
    });
  });

  it('Should update chain length when new non-pdb data is provided.', () => {
    const expected = 56;
    const wrapper = shallow(<PredictedContactMap data={emptyData} />);
    expect(wrapper.state().numPredictionsToShow).not.toBe(expected);
    wrapper.setProps({
      data: sampleData,
    });
    expect(wrapper.state().chainLength).toBe(expected);
  });

  it('Should update chain length when new pdb data is provided.', async () => {
    const expected = 56;
    const wrapper = shallow(<PredictedContactMap data={emptyData} />);
    expect(wrapper.state().numPredictionsToShow).not.toBe(expected);
    wrapper.setProps({
      data: await sampleDataWithPDB(),
    });
    expect(wrapper.state().chainLength).toBe(expected);
  });

  it('Should update number of predictions to show when new data is provided.', () => {
    const expected = 28;
    const wrapper = shallow(<PredictedContactMap data={emptyData} />);
    expect(wrapper.state().numPredictionsToShow).not.toBe(expected);
    wrapper.setProps({
      data: sampleData,
    });
    expect(wrapper.state().numPredictionsToShow).toBe(expected);
  });

  it('Should update number of predictions to show when new value is received.', () => {
    const expected = 28;
    const wrapper = shallow(<PredictedContactMap data={emptyData} />);
    expect(wrapper.state().numPredictionsToShow).not.toBe(expected);
    wrapper.setState({
      numPredictionsToShow: expected,
    });
    expect(wrapper.state().numPredictionsToShow).toBe(expected);
  });

  it('Should update points to plot when new data is provided.', () => {
    const wrapper = shallow(<PredictedContactMap data={emptyData} />);
    expect(wrapper.state().pointsToPlot).toEqual([]);
    wrapper.setProps({
      data: sampleData,
    });
    expect(wrapper.state().pointsToPlot).not.toEqual([]);
    expect(wrapper.state().pointsToPlot).toMatchSnapshot();
  });

  it('Should update points to plot using closest atom for distance determination.', async () => {
    const wrapper = shallow(<PredictedContactMap data={emptyData} />);
    expect(wrapper.state().pointsToPlot).toEqual([]);
    wrapper.setProps({
      data: await sampleDataWithPDB(),
    });
    wrapper.setState({
      measuredProximity: CONTACT_DISTANCE_PROXIMITY.CLOSEST,
    });
    expect(wrapper.state().pointsToPlot).not.toEqual([]);
    expect(wrapper.state().pointsToPlot).toMatchSnapshot();
  });

  it('Should update points to plot using C-Alpha for distance determination.', async () => {
    const wrapper = shallow(<PredictedContactMap data={emptyData} />);
    expect(wrapper.state().pointsToPlot).toEqual([]);
    wrapper.setProps({
      data: await sampleDataWithPDB(),
    });
    wrapper.setState({
      measuredProximity: CONTACT_DISTANCE_PROXIMITY.C_ALPHA,
    });
    expect(wrapper.state().pointsToPlot).not.toEqual([]);
    expect(wrapper.state().pointsToPlot).toMatchSnapshot();
  });
});
