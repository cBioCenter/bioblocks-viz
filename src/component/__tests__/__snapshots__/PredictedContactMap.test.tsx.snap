// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PredictedContactMap Should update points to plot using C-Alpha for distance determination. 1`] = `
Array [
  Object {
    "hoverinfo": "x+y",
    "marker": Object {
      "colorscale": Array [
        Array [
          0,
          "rgb(12,50,102)",
        ],
        Array [
          0.1,
          "rgb(17,83,150)",
        ],
        Array [
          0.2,
          "rgb(40,114,175)",
        ],
        Array [
          0.3,
          "rgb(71,147,193)",
        ],
        Array [
          0.4,
          "rgb(111,175,209)",
        ],
        Array [
          0.5,
          "rgb(160,202,222)",
        ],
        Array [
          0.6,
          "rgb(200,219,237)",
        ],
        Array [
          0.7,
          "rgb(224,235,246)",
        ],
        Array [
          0.8,
          "rgb(247,251,255)",
        ],
        Array [
          0.9,
          "rgb(249,253,255)",
        ],
        Array [
          1,
          "rgb(255,255,255)",
        ],
      ],
    },
    "mode": "lines+markers",
    "name": "Known Structure Contact",
    "nodeSize": 4,
    "points": Array [
      Object {
        "dist": 1.3,
        "i": 41,
        "j": 52,
      },
      Object {
        "dist": 2.4,
        "i": 56,
        "j": 50,
      },
    ],
    "subtitle": "(from PDB structure)",
  },
  Object {
    "hoverinfo": "x+y",
    "marker": Object {
      "color": "#000000",
    },
    "mode": "lines+markers",
    "name": "Predicted Contact",
    "nodeSize": 4,
    "points": Array [
      Object {
        "dist": 1.3,
        "i": 41,
        "j": 52,
      },
      Object {
        "dist": 20.4,
        "i": 42,
        "j": 50,
      },
      Object {
        "dist": 2.4,
        "i": 56,
        "j": 50,
      },
    ],
    "subtitle": "(N=28, L=56)",
  },
  Object {
    "hoverinfo": "x+y",
    "marker": Object {
      "color": "#ff0000",
    },
    "mode": "lines+markers",
    "name": "Correct Prediction",
    "nodeSize": 6,
    "points": Array [
      Object {
        "dist": 1.3,
        "i": 41,
        "j": 52,
      },
      Object {
        "dist": 2.4,
        "i": 56,
        "j": 50,
      },
    ],
    "subtitle": "(N=2, 66.67%)",
  },
]
`;

exports[`PredictedContactMap Should update points to plot using closest atom for distance determination. 1`] = `
Array [
  Object {
    "hoverinfo": "x+y",
    "marker": Object {
      "colorscale": Array [
        Array [
          0,
          "rgb(12,50,102)",
        ],
        Array [
          0.1,
          "rgb(17,83,150)",
        ],
        Array [
          0.2,
          "rgb(40,114,175)",
        ],
        Array [
          0.3,
          "rgb(71,147,193)",
        ],
        Array [
          0.4,
          "rgb(111,175,209)",
        ],
        Array [
          0.5,
          "rgb(160,202,222)",
        ],
        Array [
          0.6,
          "rgb(200,219,237)",
        ],
        Array [
          0.7,
          "rgb(224,235,246)",
        ],
        Array [
          0.8,
          "rgb(247,251,255)",
        ],
        Array [
          0.9,
          "rgb(249,253,255)",
        ],
        Array [
          1,
          "rgb(255,255,255)",
        ],
      ],
    },
    "mode": "lines+markers",
    "name": "Known Structure Contact",
    "nodeSize": 4,
    "points": Array [
      Object {
        "dist": 1.3,
        "i": 41,
        "j": 52,
      },
      Object {
        "dist": 2.4,
        "i": 56,
        "j": 50,
      },
    ],
    "subtitle": "(from PDB structure)",
  },
  Object {
    "hoverinfo": "x+y",
    "marker": Object {
      "color": "#000000",
    },
    "mode": "lines+markers",
    "name": "Predicted Contact",
    "nodeSize": 4,
    "points": Array [
      Object {
        "dist": 1.3,
        "i": 41,
        "j": 52,
      },
      Object {
        "dist": 20.4,
        "i": 42,
        "j": 50,
      },
    ],
    "subtitle": "(N=-1, L=56)",
  },
  Object {
    "hoverinfo": "x+y",
    "marker": Object {
      "color": "#ff0000",
    },
    "mode": "lines+markers",
    "name": "Correct Prediction",
    "nodeSize": 6,
    "points": Array [
      Object {
        "dist": 1.3,
        "i": 41,
        "j": 52,
      },
    ],
    "subtitle": "(N=1, 50.00%)",
  },
]
`;

exports[`PredictedContactMap Should update points to plot when new data is provided. 1`] = `
Array [
  Object {
    "hoverinfo": "x+y",
    "marker": Object {
      "colorscale": Array [
        Array [
          0,
          "rgb(12,50,102)",
        ],
        Array [
          0.1,
          "rgb(17,83,150)",
        ],
        Array [
          0.2,
          "rgb(40,114,175)",
        ],
        Array [
          0.3,
          "rgb(71,147,193)",
        ],
        Array [
          0.4,
          "rgb(111,175,209)",
        ],
        Array [
          0.5,
          "rgb(160,202,222)",
        ],
        Array [
          0.6,
          "rgb(200,219,237)",
        ],
        Array [
          0.7,
          "rgb(224,235,246)",
        ],
        Array [
          0.8,
          "rgb(247,251,255)",
        ],
        Array [
          0.9,
          "rgb(249,253,255)",
        ],
        Array [
          1,
          "rgb(255,255,255)",
        ],
      ],
    },
    "mode": "lines+markers",
    "name": "Known Structure Contact",
    "nodeSize": 4,
    "points": Array [
      Object {
        "dist": 1.3,
        "i": 41,
        "j": 52,
      },
      Object {
        "dist": 2.4,
        "i": 56,
        "j": 50,
      },
    ],
    "subtitle": "(from PDB structure)",
  },
  Object {
    "hoverinfo": "x+y",
    "marker": Object {
      "color": "#000000",
    },
    "mode": "lines+markers",
    "name": "Predicted Contact",
    "nodeSize": 4,
    "points": Array [
      Object {
        "dist": 1.3,
        "i": 41,
        "j": 52,
      },
      Object {
        "dist": 20.4,
        "i": 42,
        "j": 50,
      },
    ],
    "subtitle": "(N=-1, L=56)",
  },
  Object {
    "hoverinfo": "x+y",
    "marker": Object {
      "color": "#ff0000",
    },
    "mode": "lines+markers",
    "name": "Correct Prediction",
    "nodeSize": 6,
    "points": Array [
      Object {
        "dist": 1.3,
        "i": 41,
        "j": 52,
      },
    ],
    "subtitle": "(N=1, 50.00%)",
  },
]
`;

exports[`PredictedContactMap Snapshots Should match existing snapshot when given empty data. 1`] = `
<div
  id="PredictedContactMapComponent"
>
  <ContactMap
    addCandidateResidues={[Function]}
    addHoveredResidues={[Function]}
    addLockedResiduePair={[Function]}
    candidateResidues={Array []}
    chainLength={-1}
    clearAllResidues={[Function]}
    configurations={
      Array [
        Object {
          "name": "Calculate Distance: Closest Atom - C-Alpha",
          "onChange": [Function],
          "type": "RADIO",
          "values": Object {
            "current": 0,
            "max": 1,
            "min": 0,
            "options": Array [
              "CLOSEST",
              "C-α",
            ],
          },
        },
        Object {
          "name": "Linear Distance Filter (|i - j|)",
          "onChange": [Function],
          "type": "SLIDER",
          "values": Object {
            "current": 5,
            "max": 10,
            "min": 1,
          },
        },
        Object {
          "name": "Top N Predictions to Show",
          "onChange": [Function],
          "type": "SLIDER",
          "values": Object {
            "current": -1,
            "max": -1,
            "min": 1,
          },
        },
      ]
    }
    correctColor="#ff0000"
    data={
      Object {
        "computedPoints": Array [],
        "secondaryStructures": Array [],
      }
    }
    height={400}
    hoveredResidues={Array []}
    incorrectColor="#000000"
    lockedResiduePairs={Map {}}
    observedColor="#0000ff"
    padding={0}
    removeAllLockedResiduePairs={[Function]}
    removeCandidateResidues={[Function]}
    removeHoveredResidues={[Function]}
    removeLockedResiduePair={[Function]}
    removeNonLockedResidues={[Function]}
    toggleLockedResiduePair={[Function]}
    width={400}
  />
</div>
`;

exports[`PredictedContactMap Snapshots Should match existing snapshot when given no data. 1`] = `
<div
  id="PredictedContactMapComponent"
>
  <ContactMap
    addCandidateResidues={[Function]}
    addHoveredResidues={[Function]}
    addLockedResiduePair={[Function]}
    candidateResidues={Array []}
    chainLength={-1}
    clearAllResidues={[Function]}
    configurations={
      Array [
        Object {
          "name": "Calculate Distance: Closest Atom - C-Alpha",
          "onChange": [Function],
          "type": "RADIO",
          "values": Object {
            "current": 0,
            "max": 1,
            "min": 0,
            "options": Array [
              "CLOSEST",
              "C-α",
            ],
          },
        },
        Object {
          "name": "Linear Distance Filter (|i - j|)",
          "onChange": [Function],
          "type": "SLIDER",
          "values": Object {
            "current": 5,
            "max": 10,
            "min": 1,
          },
        },
        Object {
          "name": "Top N Predictions to Show",
          "onChange": [Function],
          "type": "SLIDER",
          "values": Object {
            "current": -1,
            "max": -1,
            "min": 1,
          },
        },
      ]
    }
    correctColor="#ff0000"
    data={
      Object {
        "computedPoints": Array [],
        "secondaryStructures": Array [],
      }
    }
    height={400}
    hoveredResidues={Array []}
    incorrectColor="#000000"
    lockedResiduePairs={Map {}}
    observedColor="#0000ff"
    padding={0}
    removeAllLockedResiduePairs={[Function]}
    removeCandidateResidues={[Function]}
    removeHoveredResidues={[Function]}
    removeLockedResiduePair={[Function]}
    removeNonLockedResidues={[Function]}
    toggleLockedResiduePair={[Function]}
    width={400}
  />
</div>
`;

exports[`PredictedContactMap Snapshots Should match snapshot when locked residues are added. 1`] = `
<div
  id="PredictedContactMapComponent"
>
  <ContactMap
    addCandidateResidues={[Function]}
    addHoveredResidues={[Function]}
    addLockedResiduePair={[Function]}
    candidateResidues={Array []}
    chainLength={-1}
    clearAllResidues={[Function]}
    configurations={
      Array [
        Object {
          "name": "Calculate Distance: Closest Atom - C-Alpha",
          "onChange": [Function],
          "type": "RADIO",
          "values": Object {
            "current": 0,
            "max": 1,
            "min": 0,
            "options": Array [
              "CLOSEST",
              "C-α",
            ],
          },
        },
        Object {
          "name": "Linear Distance Filter (|i - j|)",
          "onChange": [Function],
          "type": "SLIDER",
          "values": Object {
            "current": 5,
            "max": 10,
            "min": 1,
          },
        },
        Object {
          "name": "Top N Predictions to Show",
          "onChange": [Function],
          "type": "SLIDER",
          "values": Object {
            "current": -1,
            "max": -1,
            "min": 1,
          },
        },
      ]
    }
    correctColor="#ff0000"
    data={
      Object {
        "computedPoints": Array [],
        "secondaryStructures": Array [],
      }
    }
    height={400}
    hoveredResidues={Array []}
    incorrectColor="#000000"
    lockedResiduePairs={
      Object {
        "37,46": Array [
          37,
          46,
        ],
        "8": Array [
          8,
        ],
      }
    }
    observedColor="#0000ff"
    padding={0}
    removeAllLockedResiduePairs={[Function]}
    removeCandidateResidues={[Function]}
    removeHoveredResidues={[Function]}
    removeLockedResiduePair={[Function]}
    removeNonLockedResidues={[Function]}
    toggleLockedResiduePair={[Function]}
    width={400}
  />
</div>
`;
